### PR TITLE
Fix an issue with multiple short list rowgroups using the Parquet chunked reader.

### PR DESCRIPTION
**Pull request by nvdbaranec**
Fixes https://github.com/rapidsai/cudf/issues/15306

The core issue here was that under certain conditions, the chunked reader could generate invalid page indices for list columns when using the chunked reader.  This led to corruption in the decode kernels.  The fix is fairly simple, but there's a decent amount of delta in this PR that is just name changes for clarity and some more comments/docs.

This affected the number of chunks generated in some of the very (unrealistically) constrained tests.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by nvdbaranec** [comment]
Leaving as a draft until running it through all the Spark tests.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 1]
```diff
@@ -318,14 +318,15 @@ size_t find_start_index(cudf::host_span<cumulative_page_info const> aggregated_i
...
-  // we find the next group of pages
-  while (start_index < (static_cast<int64_t>(aggregated_info.size()) - 1) &&
-         (start_index < 0 || aggregated_info[start_index].row_index == start_row)) {
+  // list rows can span page boundaries, so it is not always safe to assume that the row
+  // represented by end_row_index starts on the subsequent page. It is possible that
+  // the values for row end_row_index start within the page itself. so we must
+  // include the page in that case.
+  while (start_index < (static_cast<int64_t>(aggregated_info.size()) - 1) && (start_index < 0)) {
     start_index++;
   }
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_chunking.cu:331]: doesn't look like we need to iterate here.
```suggestion
  start_index = std::min(aggregated_info.size() - 1, 0);

```

Comment (vuule) [reply cpp/src/io/parquet/reader_impl_chunking.cu:331]: from offline discussion: this can be removed, start_index is >=0 in the first place.

**Comment by nvdbaranec** [comment]
Adding do-not-merge until pending changes run through the spark integration tests.

**Review by vuule — COMMENTED** [review]
few small suggestions, looks good otherwise
tricky stuff

File: cpp/src/io/parquet/reader_impl_chunking.cu [hunk 2]
```diff
@@ -48,9 +48,9 @@ struct split_info {
 };
 
 struct cumulative_page_info {
-  size_t row_index;   // row index
-  size_t size_bytes;  // cumulative size in bytes
-  int key;            // schema index
+  size_t end_row_index;  // end row index (exclusive)
```
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/src/io/parquet/reader_impl_chunking.cu:51]: That's not what `end_row_index` represents.  If you had a page that starts on row 0 and contains 1 row, `end_row_index` would be 1.  But only row 0 ends on the page.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/parquet/reader_impl_chunking.cu:51]: somehow I thought that "(exclusive)" accounts for this.
I guess it's either 
"index of the last row that ends in this page, plus one"
"index of the first row that ends after this page"
Whatever wording, as long as it clarifies that it's about the end of the row.

**Review by vuule — APPROVED** [review]
Thank you for addressing the suggestions!

**Pull request merged**